File: 0143-reorder-list/0143-reorder-list.rb
# Definition for singly-linked list.
# class ListNode
#     attr_accessor :val, :next
#     def initialize(val = 0, _next = nil)
#         @val = val
#         @next = _next
#     end
# end
# @param {ListNode} head
# @return {Void} Do not return anything, modify head in-place instead.
def reorder_list(head)
    arr = []
    node = head
    
    while node
        arr << node
        node = node.next
    end
    
    l, r = 1, arr.length - 1
    
    while l <= r
        head.next = arr[r]
        r -= 1
        head = head.next
        head.next = arr[l]
        l += 1
        head = head.next
    end
    
    head.next = nil
end

File: 0143-reorder-list/0143-reorder-list.py
from collections import deque
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        """
        Do not return anything, modify head in-place instead.
        """
        stack = deque([])
        tail = head.next

        while tail:
            stack.append(tail)
            tail = tail.next

        tail = head

        while stack:
            if stack:
                popped = stack.pop()
                popped.next = None
                tail.next = popped
                tail = tail.next
            if stack:
                popped = stack.popleft()
                popped.next = None
                tail.next = popped
                tail = tail.next
        
        return head
        